MIC PODs crashing - after attempting to update user-assigned identities, panic: runtime error: invalid memory address or nil pointer dereference

Hi All,

I'm facing an issue with geting aad-pod-identity to work and I am uncertain what the issue is.
I have tried similar steps in a different AKS cluster and it is running fine, although the AKS configuration is different there, but I dont know what the difference is or what is causing this issue. Maybe it is a permissions issue?

I installed it via Helm (v3) charts, without any variables, so I assume it should run without issues.

```
The MIC PODS are in CrashLoopBackOff status, with this as the typical log:
I0205 09:15:06.857150       1 main.go:114] starting mic process. Version: v1.7.2. Build date: 2021-02-02-21:59
W0205 09:15:06.857218       1 main.go:119] --kubeconfig not passed will use InClusterConfig
I0205 09:15:06.857226       1 main.go:136] kubeconfig () cloudconfig (/etc/kubernetes/azure.json)
I0205 09:15:06.857374       1 main.go:144] running MIC in namespaced mode: false
I0205 09:15:06.857384       1 main.go:148] client QPS set to: 5. Burst to: 5
I0205 09:15:06.857399       1 mic.go:139] starting to create the pod identity client. Version: v1.7.2. Build date: 2021-02-02-21:59
I0205 09:15:06.880587       1 mic.go:145] Kubernetes server version: v1.18.10
I0205 09:15:06.880869       1 cloudprovider.go:123] MIC using user assigned identity: 6b2e##### REDACTED #####8ac2 for authentication.
I0205 09:15:06.946516       1 probes.go:41] initialized health probe on port 8080
I0205 09:15:06.946534       1 probes.go:44] started health probe
I0205 09:15:06.946648       1 metrics.go:341] registered views for metric
I0205 09:15:06.946696       1 prometheus_exporter.go:21] starting Prometheus exporter
I0205 09:15:06.946702       1 metrics.go:347] registered and exported metrics on port 8888
I0205 09:15:06.946707       1 mic.go:240] initiating MIC Leader election
I0205 09:15:06.946716       1 leaderelection.go:243] attempting to acquire leader lease  default/aad-pod-identity-mic...
I0205 09:15:06.979031       1 leaderelection.go:253] successfully acquired lease default/aad-pod-identity-mic
I0205 09:15:06.987329       1 mic.go:328] type upgrade status configmap found from version: v1.7.2. Skipping type upgrade!
I0205 09:15:07.088170       1 crd.go:456] CRD informers started
I0205 09:15:07.487568       1 pod.go:72] pod cache synchronized. Took 500.176071ms
I0205 09:15:07.487619       1 pod.go:79] pod watcher started !!
I0205 09:15:07.587706       1 mic.go:398] sync thread started.
I0205 09:15:07.587896       1 mic.go:1023] processing node aks-lnx001-22602557-vmss, add [0], del [4], update [0]
I0205 09:15:07.900002       1 cloudprovider.go:210] updating user-assigned identities on aks-lnx001-22602557-vmss, assign [0], unassign [0]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xe17c88]

goroutine 108 [running]:
github.com/Azure/aad-pod-identity/pkg/cloudprovider.(*VMSSClient).UpdateIdentities(0xc0001ae000, 0xc0004ad820, 0x17, 0xc00060469d, 0x18, 0x0, 0x0, 0x0, 0x0, 0xc0004670e0, ...)
        /go/src/github.com/Azure/aad-pod-identity/pkg/cloudprovider/vmss.go:65 +0x68
github.com/Azure/aad-pod-identity/pkg/cloudprovider.(*Client).getIdentityResource.func1(0x36b3d111daa0e, 0x35a64aee)
        /go/src/github.com/Azure/aad-pod-identity/pkg/cloudprovider/cloudprovider.go:256 +0xd5
github.com/Azure/aad-pod-identity/pkg/retry.(*client).Do(0xc00015cb80, 0xc00057ccc0, 0xc0009b8c60, 0x0, 0x18c2fc6)
        /go/src/github.com/Azure/aad-pod-identity/pkg/retry/retry.go:49 +0x2b
github.com/Azure/aad-pod-identity/pkg/cloudprovider.(*Client).UpdateUserMSI(0xc000560340, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00060469d, 0x18, 0x1, ...)
        /go/src/github.com/Azure/aad-pod-identity/pkg/cloudprovider/cloudprovider.go:237 +0x3de
github.com/Azure/aad-pod-identity/pkg/mic.(*Client).updateUserMSI(0xc0001621c0, 0xc00083ce70, 0xc00060469d, 0x18, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/Azure/aad-pod-identity/pkg/mic/mic.go:1092 +0x8cd
created by github.com/Azure/aad-pod-identity/pkg/mic.(*Client).updateNodeAndDeps
        /go/src/github.com/Azure/aad-pod-identity/pkg/mic/mic.go:1016 +0x19b

```


The NMI Pods are running, and these are the typical logs:

```

I0205 09:14:22.482146       1 main.go:84] starting nmi process. Version: v1.7.2. Build date: 2021-02-02-22:00.
I0205 09:14:22.586319       1 crd.go:447] CRD lite informers started
I0205 09:14:22.686471       1 main.go:111] running NMI in namespaced mode: false
I0205 09:14:22.686496       1 nmi.go:53] initializing in standard mode
I0205 09:14:22.686519       1 probes.go:41] initialized health probe on port 8085
I0205 09:14:22.686526       1 probes.go:44] started health probe
I0205 09:14:22.686606       1 metrics.go:341] registered views for metric
I0205 09:14:22.686703       1 prometheus_exporter.go:21] starting Prometheus exporter
I0205 09:14:22.686740       1 metrics.go:347] registered and exported metrics on port 9090
I0205 09:14:22.686750       1 server.go:98] listening on port 2579
W0205 09:14:22.799309       1 iptables.go:123] flushing iptables to add aad-metadata custom chains
I0205 09:15:14.858931       1 server.go:358] exception pod kube-system/omsagent-dz9pn token handling
I0205 09:15:14.858951       1 server.go:301] fetching token for user assigned MSI for resource: https://monitoring.azure.com/
I0205 09:15:14.869251       1 server.go:192] status (200) took 10380650 ns for req.method=GET reg.path=/metadata/identity/oauth2/token req.remote=10.192.158.10
I0205 09:15:27.823015       1 server.go:358] exception pod default/aad-pod-identity-mic-d86975cbb-pbff5 token handling
I0205 09:15:27.823064       1 server.go:301] fetching token for user assigned MSI for resource: https://management.azure.com/
I0205 09:15:27.832405       1 server.go:192] status (200) took 9452646 ns for req.method=GET reg.path=/metadata/identity/oauth2/token req.remote=10.192.158.20
I0205 09:15:31.133695       1 server.go:358] exception pod default/aad-pod-identity-mic-d86975cbb-pbff5 token handling
I0205 09:15:31.133715       1 server.go:301] fetching token for user assigned MSI for resource: https://management.azure.com/
I0205 09:15:31.143359       1 server.go:192] status (200) took 9700147 ns for req.method=GET reg.path=/metadata/identity/oauth2/token req.remote=10.192.158.20
I0205 09:15:50.090195       1 server.go:358] exception pod default/aad-pod-identity-mic-d86975cbb-pbff5 token handling
I0205 09:15:50.090212       1 server.go:301] fetching token for user assigned MSI for resource: https://management.azure.com/
I0205 09:15:50.098839       1 server.go:192] status (200) took 8699242 ns for req.method=GET reg.path=/metadata/identity/oauth2/token req.remote=10.192.158.20
I0205 09:16:06.381112       1 server.go:358] exception pod kube-system/azure-policy-765d9d7cbf-gzcch token handling
I0205 09:16:06.381131       1 server.go:301] fetching token for user assigned MSI for resource: 8cae6e77-e04e-42ce-b5cb-50d82bce26b1
I0205 09:16:06.389640       1 server.go:192] status (200) took 8545842 ns for req.method=GET reg.path=/metadata/identity/oauth2/token req.remote=10.192.158.13
I0205 09:16:43.235103       1 server.go:358] exception pod default/aad-pod-identity-mic-d86975cbb-pbff5 token handling
I0205 09:16:43.235124       1 server.go:301] fetching token for user assigned MSI for resource: https://management.azure.com/
I0205 09:16:43.248677       1 server.go:192] status (200) took 13631866 ns for req.method=GET reg.path=/metadata/identity/oauth2/token req.remote=10.192.158.20
I0205 09:17:48.308695       1 server.go:358] exception pod default/aad-pod-identity-mic-d86975cbb-pbff5 token handling
I0205 09:17:48.308738       1 server.go:301] fetching token for user assigned MSI for resource: https://management.azure.com/
I0205 09:17:48.384775       1 server.go:192] status (200) took 76112970 ns for req.method=GET reg.path=/metadata/identity/oauth2/token req.remote=10.192.158.20
I0205 09:19:40.296820       1 server.go:358] exception pod default/aad-pod-identity-mic-d86975cbb-pbff5 token handling
I0205 09:19:40.297106       1 server.go:301] fetching token for user assigned MSI for resource: https://management.azure.com/
I0205 09:19:40.307489       1 server.go:192] status (200) took 10716252 ns for req.method=GET reg.path=/metadata/identity/oauth2/token req.remote=10.192.158.20

```

**Steps To Reproduce**
I'm unsure if this is easily recreatable, I'm not sure what it is about my AKS cluster that may be the issue.

Provision AKS cluster with RBAC enabled
Install aad-pod-identity using HelmChart - default install


**Expected behavior**
Successful running aad-pod-identity

**AAD Pod Identity version**
v1.7.2. Build date: 2021-02-02-22:00.

**Kubernetes version**
1.18.10

**Additional context**

--- a/pkg/cloudprovider/vm.go
+++ b/pkg/cloudprovider/vm.go
@@ -99,7 +99,8 @@ func (c *VMClient) Get(rgName string, nodeName string) (compute.VirtualMachine, 
 
 // UpdateIdentities updates the user assigned identities for the provided node
 func (c *VMClient) UpdateIdentities(rg, nodeName string, vm compute.VirtualMachine) error {
-	if *vm.ProvisioningState == string(compute.ProvisioningStateDeleting) {
+	// if provisioning state is nil, we keep backward compatibility and proceed with the operation
+	if vm.ProvisioningState != nil && *vm.ProvisioningState == string(compute.ProvisioningStateDeleting) {
 		return fmt.Errorf("failed to update identities for %s in %s, vm is in '%s' provisioning state", nodeName, rg, *vm.ProvisioningState)
 	}
 

--- a/pkg/cloudprovider/vmss.go
+++ b/pkg/cloudprovider/vmss.go
@@ -62,7 +62,8 @@ func NewVMSSClient(config config.AzureConfig, spt *adal.ServicePrincipalToken) (
 
 // UpdateIdentities updates the user assigned identities for the provided node
 func (c *VMSSClient) UpdateIdentities(rg, vmssName string, vmss compute.VirtualMachineScaleSet) error {
-	if *vmss.ProvisioningState == string(compute.ProvisioningStateDeleting) {
+	// if provisioning state is nil, we keep backward compatibility and proceed with the operation
+	if vmss.ProvisioningState != nil && *vmss.ProvisioningState == string(compute.ProvisioningStateDeleting) {
 		return fmt.Errorf("failed to update identities for %s in %s, vmss is in %s provisioning state", vmssName, rg, *vmss.ProvisioningState)
 	}
 

